SimCalorimetry/HGCalAssociatorProducers/plugins/LayerClusterAssociatorByEnergyScoreImpl.cc

```diff
@@ -0,0 +1,531 @@
+        if (!hitWithNoCP) {
+          auto findHitIt = std::find(detIdToCaloParticleId_Map[rh_detid].begin(),
+                                     detIdToCaloParticleId_Map[rh_detid].end(),
+                                     hgcal::detIdInfoInCluster{cpPair.first, 0.f});
+          if (findHitIt != detIdToCaloParticleId_Map[rh_detid].end())
+            cpFraction = findHitIt->fraction;
+        }
+        cpPair.second +=
+            (rhFraction - cpFraction) * (rhFraction - cpFraction) * hitEnergyWeight * invLayerClusterEnergyWeight;
+      }
+    }  // End of loop over Hits within a LayerCluster
+#ifdef MRDEBUG
+    if (cpsInLayerCluster[lcId].empty())
+      LogDebug("LayerClusterAssociatorByEnergyScoreImpl") << "layerCluster Id: \t" << lcId << "\tCP id:\t-1 "
+                                                          << "\t score \t-1"
+                                                          << "\n";
+#endif
+  }  // End of loop over LayerClusters
+
+  // Compute the CaloParticle-To-LayerCluster score
+  for (const auto& cpId : cPIndices) {
+    for (unsigned int layerId = 0; layerId < layers_ * 2; ++layerId) {
+      float CPenergy = cPOnLayer[cpId][layerId].energy;
+      unsigned int CPNumberOfHits = cPOnLayer[cpId][layerId].hits_and_fractions.size();
+      if (CPNumberOfHits == 0)
+        continue;
+#ifdef MRDEBUG
+      int lcWithMaxEnergyInCP = -1;
+      float maxEnergyLCinCP = 0.f;
+      float CPEnergyFractionInLC = 0.f;
+      for (auto& lc : cPOnLayer[cpId][layerId].layerClusterIdToEnergyAndScore) {
+        if (lc.second.first > maxEnergyLCinCP) {
+          maxEnergyLCinCP = lc.second.first;
+          lcWithMaxEnergyInCP = lc.first;
+        }
+      }
+      if (CPenergy > 0.f)
+        CPEnergyFractionInLC = maxEnergyLCinCP / CPenergy;
+
+      LogDebug("LayerClusterAssociatorByEnergyScoreImpl")
+          << std::setw(8) << "LayerId:\t" << std::setw(12) << "caloparticle\t" << std::setw(15) << "cp total energy\t"
+          << std::setw(15) << "cpEnergyOnLayer\t" << std::setw(14) << "CPNhitsOnLayer\t" << std::setw(18)
+          << "lcWithMaxEnergyInCP\t" << std::setw(15) << "maxEnergyLCinCP\t" << std::setw(20) << "CPEnergyFractionInLC"
+          << "\n";
+      LogDebug("LayerClusterAssociatorByEnergyScoreImpl")
+          << std::setw(8) << layerId << "\t" << std::setw(12) << cpId << "\t" << std::setw(15)
+          << caloParticles[cpId].energy() << "\t" << std::setw(15) << CPenergy << "\t" << std::setw(14)
+          << CPNumberOfHits << "\t" << std::setw(18) << lcWithMaxEnergyInCP << "\t" << std::setw(15) << maxEnergyLCinCP
+          << "\t" << std::setw(20) << CPEnergyFractionInLC << "\n";
+#endif
+      // Compute the correct normalization
+      float invCPEnergyWeight = 0.f;
+      for (auto const& haf : cPOnLayer[cpId][layerId].hits_and_fractions) {
+        invCPEnergyWeight +=
+            (haf.second * hitMap_->at(haf.first)->energy()) * (haf.second * hitMap_->at(haf.first)->energy());
```

Comment:
use `std::pow(...,2)` to reduce repetition (gets compiled to the same expression)